dsl_parser/functions.py

```diff
@@ -147,25 +147,83 @@ def evaluate_runtime(self, storage):
 class GetInput(Function):
 
     def __init__(self, args, **kwargs):
-        self.input_name = None
+        self.input_value = None
         super(GetInput, self).__init__(args, **kwargs)
 
     def parse_args(self, args):
-        valid_args_type = isinstance(args, basestring)
-        if not valid_args_type:
+        if not isinstance(args, basestring) \
+                and not (isinstance(args, list) and len(args) >= 1):
             raise ValueError(
-                "get_input function argument should be a string, "
-                "but is '{0}'.".format(args))
-        self.input_name = args
+                "Illegal argument(s) passed to {0} function. Expected either"
+                "a string or a list [input_name [, key1/index1 [,...]]] but "
+                "got {1}".format(self.name, args))
+        self.input_value = args
 
     def validate(self, plan):
-        if self.input_name not in plan.inputs:
+        if isinstance(self.input_value, list):
+            if self.input_value[0] not in plan.inputs:
+                raise exceptions.UnknownInputError(
+                    "get_input function references an "
+                    "unknown input '{0}' in the function args '{1}'.".format(
+                        self.input_value[0], self.input_value))
+        elif self.input_value not in plan.inputs:
             raise exceptions.UnknownInputError(
                 "get_input function references an "
-                "unknown input '{0}'.".format(self.input_name))
+                "unknown input '{0}'.".format(self.input_value))
 
     def evaluate(self, plan):
-        return plan.inputs[self.input_name]
+        if isinstance(self.input_value, list):
+            return self._get_input_attribute(plan.inputs[self.input_value[0]])
+        return plan.inputs[self.input_value]
+
+    def _get_input_attribute(self, root):
+        def _get_error_string(attr_list):
+            if not attr_list:
+                return ''
+            s = str(attr_list[0])
+            for attr in attr_list[1:]:
+                if isinstance(attr, int):
+                    s += '[{0}]'.format(attr)
+                else:
+                    s += '.{0}'.format(attr)
+            return s
+
+        if isinstance(root, dict) and 'default' in root:
```

Comment:
Do you actually get to see the dict by this point in the process? I would've expected the DSL parser to convert it beforehand?

Reply:
No, as far as I understand, the input might be something else rather than a dict.
It'll always be a dict when it's a blueprint obj, once an input is provided, it can be anything.

Reply:
Yes, but by the time get_input is called won't it always have ceased being the type of dict where we care about the 'default' key?

e.g. if the input is { 'default': 'mydefault', 'description': 'Something' } then by the time your code is executed (and assuming no actual input provided), you will receive 'mydefault'.

If that's correct then this code will only trigger in cases where it probably shouldn't (when a user has entered a dict as the input with a 'default' key in it).

Reply:
You're completely right. I think I added this in the early stages and never deleted it. This piece of code does indeed only get called whenever it shouldn't.
Fixed.